Vulkan command buffer and pool.

AshkanAliabadi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43782 WIP - experimentation
* **#42930 Vulkan command buffer and pool.**
* #42709 Vulkan resource cache.
* #42642 Vulkan descriptor and descriptor layout cache.
* #42395 Vulkan pipeline and pipeline layout cache.
* #42786 Vulkan memory allocator.
* #42325 Vulkan (source and binary) shader and shader layout cache.
* #42394 Generic Vulkan object cache.
* #42503 Move torch/csrc/utils/hash.h to c10/util/hash.h.
* #43628 Revert "Revert D23252335: Refactor Vulkan context into its own files. Use RAII."

Differential Revision: [D23252333](https://our.internmc.facebook.com/intern/diff/D23252333)

Commit:
Update on "Vulkan command buffer and pool."




[ghstack-poisoned]

Commit:
Update on "Vulkan command buffer and pool."




[ghstack-poisoned]

Commit:
Update on "Vulkan command buffer and pool."


Differential Revision: [D23252333](https://our.internmc.facebook.com/intern/diff/D23252333)

[ghstack-poisoned]

Commit:
Update on "Vulkan command buffer and pool."


Differential Revision: [D23252333](https://our.internmc.facebook.com/intern/diff/D23252333)

[ghstack-poisoned]

Commit:
Update on "Vulkan command buffer and pool."


Differential Revision: [D23252333](https://our.internmc.facebook.com/intern/diff/D23252333)

[ghstack-poisoned]

Commit:
Update on "Vulkan command buffer and pool."


Differential Revision: [D23252333](https://our.internmc.facebook.com/intern/diff/D23252333)

[ghstack-poisoned]

Commit:
Update on "Vulkan command buffer and pool."


Differential Revision: [D23252333](https://our.internmc.facebook.com/intern/diff/D23252333)

[ghstack-poisoned]

Commit:
Update on "Vulkan command buffer and pool."


Differential Revision: [D23252333](https://our.internmc.facebook.com/intern/diff/D23252333)

[ghstack-poisoned]

IvanKobzarev commented on aten/src/ATen/native/vulkan/api/Command.h:
```diff
@@ -0,0 +1,104 @@
+inline bool operator==(
+    const Command::Pool::Descriptor& _1,
+    const Command::Pool::Descriptor& _2) {
+  return _1.queue_family_index == _2.queue_family_index;
```
Never seen that kind of naming, looks cool :)

Commit:
Update on "Vulkan command buffer and pool."


Differential Revision: [D23252333](https://our.internmc.facebook.com/intern/diff/D23252333)

[ghstack-poisoned]

Commit:
Update on "Vulkan command buffer and pool."


Differential Revision: [D23252333](https://our.internmc.facebook.com/intern/diff/D23252333)

[ghstack-poisoned]

Commit:
Update on "Vulkan command buffer and pool."


Differential Revision: [D23252333](https://our.internmc.facebook.com/intern/diff/D23252333)

[ghstack-poisoned]

Commit:
Update on "Vulkan command buffer and pool."


Differential Revision: [D23252333](https://our.internmc.facebook.com/intern/diff/D23252333)

[ghstack-poisoned]
